Sources/MongoSwift/MongoCollection+FindAndModify.swift

```diff
@@ -147,4 +212,100 @@ public struct FindOneAndUpdateOptions: Encodable {
+        self.sort = sort
+        self.upsert = upsert
+        self.writeConcern = writeConcern
+    }
+}
+
+/// A class wrapping a `mongoc_find_and_modify_opts_t`, for use with `MongoCollection.findAndModify`
+private class FindAndModifyOptions {
+    // an `OpaquePointer` to a `mongoc_find_and_modify_opts_t`
+    var _options: OpaquePointer?
+
+    init() {
+        self._options = mongoc_find_and_modify_opts_new()
+    }
+
+    // swiftlint:disable:next cyclomatic_complexity
+    init(arrayFilters: [Document]? = nil, bypassDocumentValidation: Bool? = nil, collation: Document?,
+         maxTimeMS: Int64?, projection: Document?, remove: Bool? = nil, returnDocument: ReturnDocument? = nil,
+         sort: Document?, upsert: Bool? = nil, writeConcern: WriteConcern?) throws {
+        self._options = mongoc_find_and_modify_opts_new()
+
+        if let bypass = bypassDocumentValidation,
+        !mongoc_find_and_modify_opts_set_bypass_document_validation(self._options, bypass) {
+            throw MongoError.invalidArgument(message: "Error setting bypassDocumentValidation to \(bypass)")
+        }
+
+        if let fields = projection, !mongoc_find_and_modify_opts_set_fields(self._options, fields.data) {
+            throw MongoError.invalidArgument(message: "Error setting fields to \(fields)")
+        }
+
+        // build a mongoc_find_and_modify_flags_t
+        var flags = MONGOC_FIND_AND_MODIFY_NONE.rawValue
+        if remove == true { flags |= MONGOC_FIND_AND_MODIFY_REMOVE.rawValue }
+        if upsert == true { flags |= MONGOC_FIND_AND_MODIFY_UPSERT.rawValue }
+        if returnDocument == .after { flags |= MONGOC_FIND_AND_MODIFY_RETURN_NEW.rawValue }
+        let mongocFlags = mongoc_find_and_modify_flags_t(rawValue: flags)
+
+        if mongocFlags != MONGOC_FIND_AND_MODIFY_NONE,
+        !mongoc_find_and_modify_opts_set_flags(self._options, mongocFlags) {
+            let remStr = String(describing: remove)
+            let upsStr = String(describing: upsert)
+            let retStr = String(describing: returnDocument)
+            throw MongoError.invalidArgument(message:
+                "Error setting flags to \(flags); remove=\(remStr), upsert=\(upsStr), returnDocument=\(retStr)")
+        }
+
+        if let sort = sort, !mongoc_find_and_modify_opts_set_sort(self._options, sort.data) {
+            throw MongoError.invalidArgument(message: "Error setting sort to \(sort)")
+        }
+
+        // build an "extra" document of fields without their own setters
+        var extra = Document()
+        if let filters = arrayFilters { extra["arrayFilters"] = filters }
+        if let coll = collation { extra["collation"] = coll }
+        // note: mongoc_find_and_modify_opts_set_max_time_ms() takes in a 
+        // uint32_t, but it should be a positive 64-bit integer. thus, we
+        // set maxTimeMS by directly appending it instead. see CDRIVER-1329
+        if let maxTime = maxTimeMS { extra["maxTimeMS"] = maxTime }
```

Comment:
should we bother to validate that this is > 0, or wait for a complaint from further down the stack?

Reply:
[`mongoc_collection_find_and_modify_with_opts()` ignores non-positive values](https://github.com/mongodb/mongo-c-driver/blob/1.11.0/src/libmongoc/src/mongoc/mongoc-collection.c#L3258) when handling the named option. That doesn't apply here, since you're merging directly into the command document. In that case, you can trust that users will receive a server error (at least on 4.0):

```
> db.foo.runCommand({ping:1,maxTimeMS:1})
{ "ok" : 1 }
> db.foo.runCommand({ping:1,maxTimeMS:0})
{ "ok" : 1 }
> db.foo.runCommand({ping:1,maxTimeMS:-5})
{
	"ok" : 0,
	"errmsg" : "maxTimeMS is out of range",
	"code" : 2,
	"codeName" : "BadValue"
}
```

Earlier server versions could very well ignore it altogether. I've not tested them.

Reply:
👍